boxoffice/views/admin_discount.py

```diff
@@ -0,0 +1,210 @@
+        current_page=data_dict['current_page']
+    )
+
+
+@app.route('/admin/o/<org>/discount_policy')
+@lastuser.requires_login
+@load_models(
+    (Organization, {'name': 'org'}, 'organization'),
+    permission='org_admin'
+    )
+@render_with({'text/html': 'index.html', 'application/json': jsonify_discount_policies}, json=True)
+@requestargs('search', ('page', int))
+def admin_discount_policies(organization, search=None, page=1):
+    results_per_page = 20
+
+    if request.is_xhr:
+        discount_policies = organization.discount_policies
+
+        if search:
+            discount_policies = discount_policies.filter(
+                DiscountPolicy.title.ilike('%{query}%'.format(query=search)))
+
+        total_policies = discount_policies.count()
+        total_pages = int(math.ceil(total_policies / results_per_page))
+        offset = (page - 1) * results_per_page
+
+        discount_policies = discount_policies.limit(results_per_page).offset(offset).all()
+
+        return dict(
+            org=organization, title=organization.title,
+            discount_policies=discount_policies,
+            total_pages=total_pages,
+            paginated=(total_policies > results_per_page),
+            current_page=page)
+    else:
+        return dict(title=organization.title)
+
+
+@app.route('/admin/o/<org>/discount_policy/new', methods=['OPTIONS', 'POST'])
+@lastuser.requires_login
+@load_models(
+    (Organization, {'name': 'org'}, 'organization'),
+    permission='org_admin'
+    )
+@xhr_only
+def admin_add_discount_policy(organization):
+    discount_policy_form = DiscountPolicyForm(request.form)
+    all_items = Item.query.join(ItemCollection).filter(
+        ItemCollection.organization == organization).options(db.load_only('id')).all()
+    discount_policy_form.items.query = all_items
+    if not discount_policy_form.validate_on_submit():
+        return api_error(message=discount_policy_form.errors, status_code=400)
+
+    discount_policy = DiscountPolicy(organization=organization)
+    if discount_policy_form.discount_type.data:
+        discount_policy.set_secret()
+    if not discount_policy_form.is_price_based.data:
+        discount_policy.items.extend(discount_policy_form.items.data)
+    else:
+        discount_price_form = DiscountPriceForm(request.form)
+        discount_price_form.item.query = all_items
+        if not discount_price_form.validate_on_submit():
+            return api_error(message=discount_price_form.errors, status_code=400)
+    discount_policy_form.populate_obj(discount_policy)
+    discount_policy.make_name()
+    db.session.add(discount_policy)
+    db.session.commit()
+
+    if discount_policy.is_price_based:
+        discount_price = Price(discount_policy=discount_policy)
+        discount_price_form.populate_obj(discount_price)
+        discount_price.make_name()
+        db.session.add(discount_price)
+        discount_policy.items.append(discount_price.item)
```

Comment:
We have a database normalisation problem here. `DiscountPolicy` and `Price` both link to `Item` and also to each other, which means they could potentially link to unrelated `Items`. One of the models should have a pre-commit validator to ensure this does not happen.